api/access.go

```diff
@@ -76,11 +81,131 @@ func Rekey(current, proposed string) error {
 		return err
 	}
 	if err := uri.Post(&respMap, string(b)); err != nil {
-		if rekey_error, present := respMap["error"]; present {
-			return errors.New(rekey_error)
+		if rekeyError, present := respMap["error"]; present {
+			return errors.New(rekeyError)
 		}
 		return err
 	}
 
 	return nil
 }
+
+//Login hits the /v2/auth/login endpoint with the given username and password
+// strings in an attempt to create an authenticated session. Returns a sessionID
+// if successful and an error otherwise. If the error is due to bad credentials,
+// the error returned will be of type ErrUnauthorized.
+func Login(username, password string) (sessionID string, user *AuthIDOutput, err error) {
+	uri, err := ShieldURI("/v2/auth/login")
+	if err != nil {
+		return
+	}
+
+	j, err := json.Marshal(struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+	}{
+		Username: username,
+		Password: password,
+	})
+
+	if err != nil {
+		panic("Could not marshal auth struct that we JUST made")
+	}
+
+	req, err := http.NewRequest("POST", uri.String(), bytes.NewReader(j))
+	if err != nil {
+		return
+	}
+
+	r, err := makeRequest(req)
+	if err != nil {
+		return
+	}
+
+	if r.StatusCode == 200 {
+		sessionID = r.Header.Get("X-Shield-Session")
+		var body []byte
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			return
+		}
+
+		user = &AuthIDOutput{}
+		err = json.Unmarshal(body, user)
+	} else {
+		err = getAPIError(r)
+	}
+
+	return
+}
+
+//Logout contacts the /v2/auth/logout endpoint in an attempt to invalidate the
+// current backend's session
+func Logout() error {
+	uri, err := ShieldURI("/v2/auth/logout")
+	if err != nil {
+		return err
+	}
+
+	return uri.Get(nil)
+}
+
+//LogoutSession contacts the /v2/auth/logout endpoint in an attempt to invalidate the
+// requested session ID
+func LogoutSession(sessionID string) error {
+	uri, err := ShieldURI("/v2/auth/logout")
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("GET", uri.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("X-Shield-Session", sessionID)
+	r, err := makeRequest(req)
+	if err != nil {
+		return err
+	}
+
+	if r.StatusCode != 200 {
+		err = getAPIError(r)
+	}
+
+	return err
+}
+
+//AuthIDOutput contains all the information from a call to /v2/auth/id
+type AuthIDOutput struct {
+	User    AuthUser     `json:"user"`
+	Tenants []AuthTenant `json:"tenants"`
```

Comment:
You ought to be able to nest these and avoid creating visible structs that are useless outside of the context of an AuthID response...